examples/step-66/step-66.cc

```diff
@@ -0,0 +1,1156 @@
+                const std::pair<unsigned int, unsigned int> &cell_range) const;
+
+    void local_compute_diagonal(
+      const MatrixFree<dim, number> &              data,
+      LinearAlgebra::distributed::Vector<number> & dst,
+      const unsigned int &                         dummy,
+      const std::pair<unsigned int, unsigned int> &cell_range) const;
+
+    Table<2, VectorizedArray<number>> nonlinear_values;
+  };
+
+
+
+  // The constructor of the JacobainOperator just calls the constructor of the
+  // base class MatrixFreeOperators::Base, which is itself derived from the
+  // Subscriptor class.
+  template <int dim, int fe_degree, typename number>
+  JacobianOperator<dim, fe_degree, number>::JacobianOperator()
+    : MatrixFreeOperators::Base<dim,
+                                LinearAlgebra::distributed::Vector<number>>()
+  {}
+
+
+
+  // The clear function resets out the table holding the values for the
+  // nonlinearity and call the clear function of the base class.
+  template <int dim, int fe_degree, typename number>
+  void JacobianOperator<dim, fe_degree, number>::clear()
+  {
+    nonlinear_values.reinit(0, 0);
+    MatrixFreeOperators::Base<dim, LinearAlgebra::distributed::Vector<number>>::
+      clear();
+  }
+
+
+
+  // The following evlauate_newton_step function is based on the
+  // evlauate_coefficient function from step-37, however, it does not evaluate a
+  // function object, it evaluates a vector representing a finite element
+  // function, namely the last Newton step needed for the Jacobian. Therefore we
+  // set up a FEEvaluateion object and evaluate the finite element function in
+  // the quadrature points with the gather_evaluate function. Remember if we
+  // only loop over cells this function is just a wrapper around functions
+  // read_dof_values and evaluate. Since we store the evaluated values of the
+  // finite element function in a table we do not have to call integrate in
+  // combination with distribute_local_to_global or integrate scatter.
+  //
+  // This will work well and in the local_apply function we can use the values
+  // stored in the table to apply the matrix-vector product, however, we can
+  // also optimize the implementation of the Jacobian at this stage. We can
+  // directly evaluate the nonlinear function std::exp(u_h^n[q]) and store these
+  // values in the table. This saves in each call of the vmult function all
+  // evaluations of the nonlinearity.
+  template <int dim, int fe_degree, typename number>
+  void JacobianOperator<dim, fe_degree, number>::evaluate_newton_step(
+    const LinearAlgebra::distributed::Vector<number> &src)
+  {
+    const unsigned int n_cells = this->data->n_macro_cells();
+    FEEvaluation<dim, fe_degree, fe_degree + 1, 1, number> phi(*this->data);
+
+    nonlinear_values.reinit(n_cells, phi.n_q_points);
+    for (unsigned int cell = 0; cell < n_cells; ++cell)
+      {
+        phi.reinit(cell);
+        phi.gather_evaluate(src, true, false);
```

Comment:
Without having read further, I assume you use homogeneous boundary conditions? In the general case you could write:
```suggestion
        phi.read_dof_values_plain(src);
        phi.evaluate(true, false);
```
assuming that the vector you pass in has its constraints already resolved. This function ignores the constraints from hanging nodes and (homogeneous) Dirichlet conditions to be similar to what `FEValues::get_function_values` does.